### ISSUE TITLE
pl.Struct cast to str yields a *wrong* result

### ISSUE THREAD

**Issue by svaningelgem** [pola-rs/polars#10271]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl


df = pl.DataFrame({"mapping": [{"a": "b"}]})

print("Dataframe casted to str >>>")
print(df.select(pl.all().cast(str)))

print("Dataframe row-by-row >>>")
for row in df:  # type: pl.Series
    print(row)
    for idx, cell in enumerate(row):
        print(idx, ">", cell)

print("Dataframe iter-by-row >>>")
for row in df.iter_rows():  # type: tuple
    print(row)
```

### Issue description

First the result:
```text
Dataframe casted to str >>>
shape: (1, 1)
┌─────────┐
│ mapping │
│ ---     │
│ str     │
╞═════════╡
│ {"b"}   │
└─────────┘
Dataframe row-by-row >>>
shape: (1,)
Series: 'mapping' [struct[1]]
[
	{"b"}
]
0 > {'a': 'b'}
Dataframe iter-by-row >>>
({'a': 'b'},)
```

As you can see I make a simple a>b dictionary. And the data "survives". However, when cast to `str` it yields a very strange result. Namely a "set" of the values. 
### Expected behavior

What I would expect is:
```text
┌──────────────┐
│ mapping      │
│ ---          │
│ str          │
╞══════════════╡
│ {"a": "b"}   │
└──────────────┘
```

The reason why this is important is because I am calculating the maximum lengths of each column when converted into a string value (ie: prepared for output). However, the result here should be 10, and yields 5 because the stringified result is faulty.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.11
Index type:          UInt32
Platform:            Windows-10-10.0.19044-SP0
Python:              3.11.3 | packaged by Anaconda, Inc. | (main, May 15 2023, 15:41:31) [MSC v.1916 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.4.0
matplotlib:          <not installed>
numpy:               1.25.0
pandas:              1.5.3
pyarrow:             11.0.0
pydantic:            1.10.8
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>

Process finished with exit code 0
```

</details>

**Comment by avimallu** [comment]
I would say this is the expected result. Think of `struct` as a  data type for a "group of multiple columns" inside a single column. In your example, the `struct` has the field name `a` and the values `b`. See another example:

```py
df = pl.DataFrame({"mapping": [{"a": "b"}, {"a": "c"}]})
df.select(pl.all().cast(str))
```

I'm not sure what your use case is, but it just *looks* like a Python `set`, but is actually the string `{"b"}`. You can see that if you just do:

```py
df = pl.DataFrame({"mapping": [{"a": "b"}]})
df.select(pl.all().cast(str))[0, 0]
```

Gives you `'{"b"}'`, without the `print`. I agree that the cast is probably not intuitive (although I'm still not sure why you're casting a `struct` to a `string` anyway), but you can see what's happening when you do:

```py
df = pl.DataFrame({"mapping": [{"a": ["b", "b", "d"], "x": ["y", "z"]}, {"a": ["c", "c", "e"], "x": ["w", "v"]}]})
print(df.select(pl.all().cast(str)))
```

So it is creating a *kind* of set, but not in the mathematical sense, but the closest string approximation of the `struct` you have in the DataFrame.

**Comment by cmdlineluser** [comment]
It seems like the cast is an attempt to "serialize to a string" e.g. as if one did `.apply(json.dumps)`?
```python
df.select(pl.col("mapping").apply(json.dumps)) # slow

# shape: (2, 1)
# ┌────────────┐
# │ mapping    │
# │ ---        │
# │ str        │
# ╞════════════╡
# │ {"a": "b"} │
# │ {"a": "c"} │
# └────────────┘
```
There are a few existing requests for some form of `Expr.to_json()` https://github.com/pola-rs/polars/issues?q=to_json

Is this what you're trying to do @svaningelgem?

**Comment by svaningelgem** [comment]
@avimallu : I know it's not really a *set*, but I didn't know how to better describe it. I know it's still a dict underneath ;-).

Now back to the other point you asked: I want to save the dataframe as a copybook (don't ask!). As there is no native format in polars for that, I have to iterate over each row, format it according to some specifier and dump it. Now to make life easier for people I want to automate this "specifier" thingie... ie: make it auto-size. For that, I would just cast it to a string and take the max length.
Within polars, this is very fast and works in nearly all circumstances. Except for the pl.Struct it didn't as the mapping is not (properly) done. As you loose the key of the mapping. I don't know if there is a native "set" type in polars, but that's what it looks like when printed.

As @cmdlineluser says, I think yes, a json-ified version would certainly help.

But I think it's also more because when printing the dataframe, you loose the information about the key.

As @avimallu 's first example:
```python
>>> print(pl.DataFrame({"mapping": [{"a": "b"}, {"a": "c"}]})) 
shape: (2, 1)
┌───────────┐
│ mapping   │
│ ---       │
│ struct[1] │
╞═══════════╡
│ {"b"}     │
│ {"c"}     │
└───────────┘
```
As you can see, there is a lot of loss of information here.

Additionally something weird happened:
```text
>>> print(pl.DataFrame({"mapping": [{"d": "b"}, {"e": "c"}]}))  
shape: (2, 1)
┌───────────┐
│ mapping   │
│ ---       │
│ struct[1] │
╞═══════════╡
│ {"b"}     │
│ {null}    │
└───────────┘
>>> df = pl.DataFrame({"mapping": [{"d": "b"}, {"e": "c"}]})   
>>> for row in df.iter_rows():
...     print(row)
... 
({'d': 'b'},)
({'d': None},)
```

Probably this is the way polars works internally by having the "key" as something like a "sub-column", where only the first is taken. Like "mapping.d", but it doesn't know "mapping.e" because it wasn't the first element.

Dunno a use-case for this, but we process a lot of logs with each line 1 json. So it might come up in the future.


Anyway, rambling here... TLDR: I think it should show -for the mapping- the keys as well in order to not loose information in the stringified/printed overview.

**Comment by avimallu** [comment]
> Additionally something weird happened:

That's because a Python `dict` is an approximation of what a `struct` is. You need to define the keys in all dictionaries you pass to have it in the `struct` column.

> TLDR: I think it should show -for the mapping- the keys as well in order to not loose information in the stringified/printed overview.

Is `struct` the right data type for your problem though? If you can condense your problem down to a handful of rows, we could try and find a solution for your use-case in Polars outside of going the slow `apply(json.dumps)` route.

**Comment by svaningelgem** [comment]
> Is `struct` the right data type for your problem though? If you can condense your problem down to a handful of rows, we could try and find a solution for your use-case in Polars outside of going the slow `apply(json.dumps)` route.

Thanks, what I would need is the maximum length of each field when I put it into a string. So in the case of a `struct` column, the json.dumps is the closest approximation I guess.
What I will do for now is convert `struct`s into a json dump, and retrieve it then as a string.

Though it'd be grateful if the output could be changed to include the keys of the struct.
```python
pl.DataFrame({'person':[
    {
        "firstName" : "John",
        "lastName"  : "Doe",
        "age"       : 23,
        "weight"    : 23,
        "height"    : 23,
    }
]})
┌─────────────────────────┐
│ person                  │
│ ---                     │
│ struct[5]               │
╞═════════════════════════╡
│ {"John","Doe",23,23,23} │
└─────────────────────────┘
```

What is each 23 :-), during data-exploration this can be quite tricky :-). Especially with more complex structs.

**Comment by cmdlineluser** [comment]
Does anybody know what would be needed to implement `.apply(json.dumps)` natively?

The `loads` equivalent uses:
https://github.com/pola-rs/polars/blob/bfabdd50639b271da45027809cf720771e3c6685/crates/polars-ops/src/chunked_array/strings/json_path.rs#L81

Does this mean if `polars_json::ndjson::serialize` existed, we could go the other way?

**Comment by wKollendorf** [comment]
I think this is related #8912
